### DISCUSSION TITLE
Support / adopt for X-65F

### DISCUSSION THREAD

**Discussion by obma** [nirenjan/x52pro-linux#30] [General]
Hi!
I came across your project during research for my Saitek X-65F (ID 06a3:0b6a Saitek PLC Pro Flight X65 Control System), which is missing an axis (mouse y) with the default driver. So I thought about adopting your kernel module for my device - what would be an absolute first time for me.
Can you give some advice on how to get into this? What tools can I use for testing, how to find out the necesarry bit-operations on the axis, ...?
Or do you even have time and motivation to include the X65F into your project? I could offer testing and provide data ;-)

Regards,
Max

**Comment by nirenjan** [comment]
Hi Max,

While I'm not yet sure I want to support the X65F as a part of this project, I'm happy to help with the driver part. For what it's worth though, the kernel module in this project is probably not the right way to write a driver, primarily because of the way it remaps the buttons. That being said, I had made a contribution to the kernel in the past to allow the default driver to send events for the Mouse-Y axis. You can see the changes [here](https://github.com/torvalds/linux/blame/master/drivers/hid/hid-quirks.c#L159).

To start off, we need some information regarding the device itself. You may need to run some of these commands with `sudo`. Please capture the outputs of the following commands:
* `lsusb -d 06a3: -v`
* `usbhid-dump -d 06a3`

We will probably need some additional info on how Linux maps the axes and buttons, but we can leave that for later.

Cheers,
Nirenjan

**Reply by obma** [reply]
Thanks for the instant reply and offer to help!!

Looking at your contribution to the kernel I understand that it would be enough to add the product id of the x65f to hid-ids.h and add a line in hid-quirks.c to get the missing axis?

```bash
$ sudo lsusb -d 06a3: -v
Bus 001 Device 008: ID 06a3:0b6a Saitek PLC Pro Flight X65 Control System
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0         8
  idVendor           0x06a3 Saitek PLC
  idProduct          0x0b6a 
  bcdDevice            1.65
  iManufacturer           1 Saitek
  iProduct                2 Pro Flight X65 Control System
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0022
    bNumInterfaces          1
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0x80
      (Bus Powered)
    MaxPower              150mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     119
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0010  1x 16 bytes
        bInterval              10
can't get device qualifier: Resource temporarily unavailable
can't get debug descriptor: Resource temporarily unavailable
Device Status:     0x0000
  (Bus Powered)
$ sudo usbhid-dump -d 06a3
001:008:000:DESCRIPTOR         1616176311.837391
 05 01 09 04 A1 01 09 01 A1 00 09 30 09 31 15 00
 26 FF 0F 75 0C 95 02 81 02 09 35 15 00 26 FF 01
 75 09 95 01 81 02 09 32 09 33 09 34 09 36 15 00
 26 FF 00 75 08 95 04 81 02 05 09 19 01 29 32 15
 00 25 01 95 32 75 01 81 02 05 01 09 39 15 01 25
 08 35 00 46 3B 01 66 14 00 75 04 95 01 81 42 05
 05 09 24 09 26 15 00 25 0F 75 04 95 02 81 02 75
 01 95 01 81 01 C0 C0
```

**Reply by nirenjan** [reply]
Yes, the quirks approach is the best. That does take a while to get into the kernel itself, so I'll see about adding a temporary driver for the X65 based on what you've given me, so that you can use the mouse Y axis. FWIW, there's no facility to support the "mouse" feature of the Saitek joysticks, not even for the X52, so this only allows you to get input from the mouse stick Y axis.

What I would like for you to do, if you're comfortable, is to rebuild your kernel, just add lines to include the quirk. Once you have built and booted into this modified kernel, you should be able to see the mouse Y axis getting reported. That test should be sufficient, so I can submit it to the kernel mailing list to include, or you can do it yourself.

**Reply by nirenjan** [reply]
In order to write a driver for a USB HID device, we first need to understand the descriptor. The first step is to retrieve the descriptor itself, which you have in the `usbhid-dump` output. By plugging that content into [this website](https://eleccelerator.com/usbdescreqparser/), we can get the parsed output as shown below:

```
0x05, 0x01,        // Usage Page (Generic Desktop Ctrls)
0x09, 0x04,        // Usage (Joystick)
0xA1, 0x01,        // Collection (Application)
0x09, 0x01,        //   Usage (Pointer)
0xA1, 0x00,        //   Collection (Physical)
0x09, 0x30,        //     Usage (X)
0x09, 0x31,        //     Usage (Y)
0x15, 0x00,        //     Logical Minimum (0)
0x26, 0xFF, 0x0F,  //     Logical Maximum (4095)
0x75, 0x0C,        //     Report Size (12)
0x95, 0x02,        //     Report Count (2)
0x81, 0x02,        //     Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x09, 0x35,        //     Usage (Rz)
0x15, 0x00,        //     Logical Minimum (0)
0x26, 0xFF, 0x01,  //     Logical Maximum (511)
0x75, 0x09,        //     Report Size (9)
0x95, 0x01,        //     Report Count (1)
0x81, 0x02,        //     Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x09, 0x32,        //     Usage (Z)
0x09, 0x33,        //     Usage (Rx)
0x09, 0x34,        //     Usage (Ry)
0x09, 0x36,        //     Usage (Slider)
0x15, 0x00,        //     Logical Minimum (0)
0x26, 0xFF, 0x00,  //     Logical Maximum (255)
0x75, 0x08,        //     Report Size (8)
0x95, 0x04,        //     Report Count (4)
0x81, 0x02,        //     Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x05, 0x09,        //     Usage Page (Button)
0x19, 0x01,        //     Usage Minimum (0x01)
0x29, 0x32,        //     Usage Maximum (0x32)
0x15, 0x00,        //     Logical Minimum (0)
0x25, 0x01,        //     Logical Maximum (1)
0x95, 0x32,        //     Report Count (50)
0x75, 0x01,        //     Report Size (1)
0x81, 0x02,        //     Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x05, 0x01,        //     Usage Page (Generic Desktop Ctrls)
0x09, 0x39,        //     Usage (Hat switch)
0x15, 0x01,        //     Logical Minimum (1)
0x25, 0x08,        //     Logical Maximum (8)
0x35, 0x00,        //     Physical Minimum (0)
0x46, 0x3B, 0x01,  //     Physical Maximum (315)
0x66, 0x14, 0x00,  //     Unit (System: English Rotation, Length: Centimeter)
0x75, 0x04,        //     Report Size (4)
0x95, 0x01,        //     Report Count (1)
0x81, 0x42,        //     Input (Data,Var,Abs,No Wrap,Linear,Preferred State,Null State)
0x05, 0x05,        //     Usage Page (Game Ctrls)
0x09, 0x24,        //     Usage (Move Right/Left)
0x09, 0x26,        //     Usage (Move Up/Down)
0x15, 0x00,        //     Logical Minimum (0)
0x25, 0x0F,        //     Logical Maximum (15)
0x75, 0x04,        //     Report Size (4)
0x95, 0x02,        //     Report Count (2)
0x81, 0x02,        //     Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x75, 0x01,        //     Report Size (1)
0x95, 0x01,        //     Report Count (1)
0x81, 0x01,        //     Input (Const,Array,Abs,No Wrap,Linear,Preferred State,No Null Position)
0xC0,              //   End Collection
0xC0,              // End Collection

// 119 bytes
```

The report sends data in little-endian order. What you should be looking for is the `Report Size` and `Report Count` fields above, with the size being in bits. The report count indicates that this current set of `Report Size` length fields corresponds to the most recent set of `Usage` fields. So, with that explanation, we can get the fields by treating the report as one giant little-endian integer.

* 12 bits - X
* 12 bits - Y
* 9 bits - Rz
* 8 bits - Z
* 8 bits - Rx
* 8 bits - Ry
* 8 bits - Slider
* 50 bits - Buttons 1 - 50
* 4 bits - Hat
* 4 bits - Move left/right (mouse X)
* 4 bits - Move up/down (mouse Y)
* 1 bit - dummy for byte alignment

**Total:** 128 bits (16 bytes)

Now that we understand how the bits correspond to the input fields, we can get the individual fields easily, e.g. `Rz` is bit 0 of byte 4 (indices start at 0, bit 0 is the least significant bit in the byte, bit 7 is most significant) and bits 7-0 of byte 3.

Hopefully, this has given some clarity into the bit operations needed.

**Reply by nirenjan** [reply]
I've added a kernel module `saitek_x65.ko` in the `x65-driver` branch. Could you give it a try and see if it works as expected? If it does, then this should tide you over until we can get the quirks fix into the mainline kernel.

**Reply by obma** [reply]
Adding the quirks approach works as expected. With your kernel module, the mouse Y works, but mouse X gets mixed up with Hat X & Y. Moving the mouse coolie along x does nothing on mouse x, but changes both hat x & y. Tested on 5.11.6-arch1-1.

**Reply by nirenjan** [reply]
I fixed the bit twiddling, it should work now. Can you check with the updated sources? Also, please check the rest of the axes and buttons and see if they work as expected.

**Reply by nirenjan** [reply]
Oh, and do you want to send the quirks fix to the kernel or would you like me to do it?

**Reply by obma** [reply]
Your driver works now with all axes and buttons.

Would be nice if you could hand in the quirks fix.

Thanks again, you're awesome!

**Comment by nirenjan** [comment]
Support for the X65F thumbstick has been merged into Linux kernel version 5.13
